exporters/pelorus/deserialization/__init__.py

```diff
@@ -79,54 +83,119 @@
+    """
+    return {_RETAIN_SOURCE_KEY: retain}
+
+
+# python has some magic to make classes work with `issubclass` and instances work with `isinstance`,
+# even if they don't actually inherit from them. Some so-called "abstract base classes" just check
+# for the presence of certain methods. Cool!
+# However, some of them don't support that, for reasons outlined in their docs (see the first section):
+# https://docs.python.org/3.11/library/collections.abc.html
+
+# thus, we need to make our own checks. These have the same downsides as in the docs,
+# but they're "good enough". "Practicality beats purity".
+
+# we need this for mappings specifically, because a kubernetes.dynamic.resource.ResourceField has
+# __getitem__, which is actually all we care about.
+
+
+def _type_has_getitem(type_: type) -> bool:
+    "See if this type has a getitem method (supports item[key])"
+    return hasattr(type_, "__getitem__")
+
+
+# region: attrs fixes
+
+# type annotations can be represented as the types themselves, or as strings.
+# attrs seems to use strings for fields in inherited classes, which can break
+# things for us. This may be a bug that has to be fixed for them.
+# Until then, here's a small alternative.
+
+
+def _is_attrs_class(cls: type) -> TypeGuard[type["attr.AttrsInstance"]]:
     """
     Return the class if it is an attrs class, else None.
 
     This exists because `attrs.has` is not a proper TypeGuard.
     """
-    return cast("type[attr.AttrsInstance]", cls) if attrs.has(cls) else None
+    return attrs.has(cls)
+
+
+class Field(NamedTuple):
+    """
+    Relevant field information for an attrs class.
+    """
+
+    name: str
+    type: type
+    default: Union[Any, Literal[NOTHING]]
+    metadata: dict[Any, Any]
+
+
+def _fields(cls: type["attr.AttrsInstance"]) -> Iterable[Field]:
+    "Get all field information for this attrs class."
+    hints = get_type_hints(cls)
+    attrs_fields = attrs.fields(cls)
+    for field in attrs_fields:
+        yield Field(
+            field.name,
+            hints[field.name],
+            default=field.default,
+            metadata=field.metadata,
+        )
 
 
+# endregion
+
 # region: generic types
 
 # generic types such as dict[str, int] have two parts:
 # the "origin" and the "args".
 # The origin in this case is `dict`, and the args are `(str, int)`.
 # This lets us inspect these types properly.
 
-# in the future, these should be made into `typing.TypeGuard`s.
-
 
 def _extract_dict_types(type_: type) -> Optional[tuple[type, type]]:
     """
     If this type annotation is a dictionary-like, extract its key and value types.
     Otherwise return None.
 
+    dictionary-like is defined as "has __getitem__ and two type arguments".
+
     >>> assert _extract_dict_types(dict[str, int]) == (str, int)
     >>> assert _extract_dict_types(list[int]) is None
     """
     origin, args = typing.get_origin(type_), typing.get_args(type_)
     if origin is None:
```

Comment:
can it be efficient to do just one `if`? with something like
```python
if all(origin is not None, _type_has_getitem(type_), len(args) == 2):
    return args[0], args[1]
return None
```

Reply:
Good point, I could do `origin is not None and _type_has_getitem(type_) and len(args) == 2`. I've never seen `all` used this way before, but... I don't dislike it, actually 🤔 

That might be for a later "cleanup" because for every code change, we have to wait for CI and whatnot.